src/queue.rs

```diff
@@ -512,14 +512,63 @@ impl<M: GuestAddressSpace> Queue<M> {
         }
     }
 
+    /// Update avail_event on the used ring with the last index in the avail ring.
+    pub fn update_avail_event(&mut self) {
+        let mem = self.mem.memory();
+        let index_addr = match mem.checked_offset(self.avail_ring, 2) {
+            Some(ret) => ret,
+            None => {
+                // TODO log address
+                warn!("Invalid offset");
+                return;
+            }
+        };
+        // Note that last_index has no invalid values
+        let last_index: u16 = match mem.read_obj::<u16>(index_addr) {
```

Comment:
Please use the `integer-atomics` feature of vm-memory instead, and `get_atomic_ref::<u16>`. You can then use normal Rust atomic loads and stores.

Reply:
Done, please help to review the memory ordering:)